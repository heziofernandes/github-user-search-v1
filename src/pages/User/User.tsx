import { useParams, useNavigate } from 'react-router-dom';
import { Button } from '../../components/Button';
import { Spinner } from '../../components/Spinner';
import { useFetch } from '../../hooks/useFetch';

import styles from './user.module.scss';

interface UserData {
  avatar_url: string;
  name: string;
  public_repos: number;
}

interface Repository {
  full_name: string;
  description: string;
  id: number;
}

export const User = () => {
  const { username } = useParams();
  const navigate = useNavigate();
  const {
    data: user,
    isFetching: isFetchingUser,
    error: userError,
  } = useFetch<UserData>(`users/${username}`);
  const {
    data: repos,
    isFetching: isFetchingRepos,
    error: reposError,
  } = useFetch<Repository[]>(`users/${username}/repos`);

  const navigateToHome = () => navigate('/');

  if (userError || reposError) {
    return (
      <div className={styles.errorContainer}>
        <h2 className={styles.errorMessage} data-testid="user-not-found">
          Ops, something went wrong ðŸ˜¢
        </h2>
        <Button customStyle={styles.button} title="Return" onClick={navigateToHome} />
      </div>
    );
  }

  return (
    <>
      {isFetchingRepos || isFetchingUser ? (
        <div className={styles.loading}>
          <Spinner />
        </div>
      ) : (
        <div className={styles.container}>
          <h1 className={styles.title} data-testid="main-title">
            Look who we found ðŸ”¥
          </h1>
          <div className={styles.userContainer}>
            <img className={styles.avatar} src={user?.avatar_url} alt="" />
            <div className={styles.userInfo}>
              <p>
                <b>Name:</b> <span data-testid="username">{user?.name}</span>
              </p>
              <p>
                <b>Repositories:</b>{' '}
                <span data-testid="repo-count">{user?.public_repos}</span>
              </p>
              <Button
                customStyle={styles.button}
                title="Return"
                onClick={navigateToHome}
              />
            </div>
          </div>
          <div className={styles.reposContainer}>
            <h1>Repositories</h1>
            <ul className={styles.repoList}>
              {repos?.map((repo) => (
                <li className={styles.repoItem} key={repo.full_name}>
                  <strong data-testid={`reponame-${repo.id}`}>{repo.full_name}</strong>
                  <p>{repo.description}</p>
                </li>
              ))}
            </ul>
          </div>
        </div>
      )}
    </>
  );
};
