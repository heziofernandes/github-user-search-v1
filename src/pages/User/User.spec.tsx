import React from 'react';
import { render, screen } from '@testing-library/react';
import mockUser from './fixture/mockUser';
import mockRepo from './fixture/mockRepo';

import { User } from './User';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  useParams: () => ({
    username: 'testUser',
  }),
}));

const returnMock = (key: string) => {
  const Obj: any = {
    'users/testUser': { data: mockUser },
    'users/testUser/repos': { data: mockRepo },
  };
  return Obj[key];
};

jest.mock('../../hooks/useFetch', () => ({
  useFetch: (url: string) => {
    return returnMock(url);
  },
}));

describe('User Component', () => {
  it('should be rendered flawlessly', () => {
    const { container } = render(<User />);
    expect(container).toMatchSnapshot();
  });

  it('should be rendered username succefully', () => {
    const { getByTestId } = render(<User />);
    const usernameElement = screen.getByTestId('username');
    expect(usernameElement).toBeInTheDocument();
    expect(getByTestId).toMatchSnapshot();
    expect(usernameElement.innerHTML).toEqual(mockUser.name);
  });

  it('should be rendered repository name succefully', () => {
    render(<User />);
    mockRepo.forEach((item, index) => {
      const repoNameElement = screen.getByTestId(`reponame-${item.id}`);
      expect(repoNameElement.innerHTML).toEqual(mockRepo[index].full_name);
    });
  });
});
