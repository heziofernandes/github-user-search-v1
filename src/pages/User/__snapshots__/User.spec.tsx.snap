// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Component should be rendered flawlessly 1`] = `
<div>
  <div>
    <h1
      data-testid="main-title"
    >
      Look who we found ðŸ”¥
    </h1>
    <div>
      <img
        alt=""
        src="https://avatars.githubusercontent.com/u/fernandes?v=4"
      />
      <div>
        <p>
          <b>
            Name:
          </b>
           
          <span
            data-testid="username"
          >
            Thiago H S Ramos
          </span>
        </p>
        <p>
          <b>
            Repositories:
          </b>
           
          <span
            data-testid="repo-count"
          >
            25
          </span>
        </p>
        <button
          class=""
          type="button"
        >
          Return
        </button>
      </div>
    </div>
    <div>
      <h1>
        Repositories
      </h1>
      <ul>
        <li>
          <strong
            data-testid="reponame-215256985"
          >
            fernandes/dasa-code-challenge
          </strong>
          <p>
            Dasa Code Challenge frontend
          </p>
        </li>
        <li>
          <strong
            data-testid="reponame-213955809"
          >
            fernandes/graphql-blog-api
          </strong>
          <p />
        </li>
        <li>
          <strong
            data-testid="reponame-294670010"
          >
            fernandes/graphql-blog-study
          </strong>
          <p>
            Course from udemy
          </p>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`User Component should be rendered username succefully 1`] = `[Function]`;
